web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts

```diff
@@ -99,34 +99,40 @@ export class AccessRequestsService {
     });
   }
 
-  async addOrRemoveKubeNamespaces(namespaceUris: KubeResourceNamespaceUri[]) {
+  async updateNamespacesForKubeCluster(
+    namespaceUris: KubeResourceNamespaceUri[],
+    kubeClusterUri: string
+  ) {
     this.setState(draftState => {
       if (draftState.pending.kind !== 'resource') {
         throw new Error('Cannot add a kube namespace to a role access request');
       }
 
       const { resources } = draftState.pending;
 
+      // Validate each namespace uri's.
       namespaceUris.forEach(resourceUri => {
-        const requestedResource = resources.get(
+        const kubeRequestedResource = resources.get(
```

Comment:
But maybe we could partially follow Grzegorz's advice, but instead of dropping namespace URIs, we could drop `kubeClusterUri` as an arg of this specific method of `AccessRequestsService`. At this layer, we just need namespace URIs alone.

Then we could convert them all to a kube cluster URI, put them in a set, check if its length is 1 (this ensures all namespace URIs point to a single cluster) and then we're pretty much set.

Reply:
> Instead, I'd implement something similar to routing.belongsToProfile, e.g., routing.belongsToKubeCluster which checks if two URIs point to something under the same kube cluster. Then you could check if each URL in the array belongs to the cluster that kubeClusterUri points to.

i wasn't sure i understood, but doesn't this (code below) already do what you suggested? (or do you mean like pull it out into its function?)

```ts
namespaceUris.forEach(namespaceUri => {
        const namespaceKubeResource = resources.get(
          routing.getKubeUri(
            routing.parseKubeResourceNamespaceUri(namespaceUri).params
          )
        );
        if (namespaceKubeResource.resource.uri != kubeClusterUri) {
          throw new Error(
            'Only namespace belonging to the same requested kube cluster can be updated'
          );
        }
      });
```

> we could drop kubeClusterUri as an arg of this specific method of AccessRequestsService

hrm... but what if a namespace was cleared? i would still need to know the kube cluster URI to get to the namespaces b/c my list of namespaces will be empty:

```ts
      const kubeRequestedResource = resources.get(kubeClusterUri);   // here
      if (kubeRequestedResource.kind === 'kube') {
        kubeRequestedResource.resource.namespaces = new Set(namespaceUris);  // namespaceUris could be empty
      }
```

Reply:
> > Instead, I'd implement something similar to routing.belongsToProfile, e.g., routing.belongsToKubeCluster which checks if two URIs point to something under the same kube cluster. Then you could check if each URL in the array belongs to the cluster that kubeClusterUri points to.
> 
> i wasn't sure i understood, but doesn't this (code below) already do what you suggested? (or do you mean like pull it out into its function?)

It does, but:

1. We take `namespaceUri` and convert it to a kube cluster URI
2. Then we `resources.get` with that URI
3. Then we look at `namespaceKubeResource.resource.uri`

Aren't the last two steps unnecessary? We can convert straight from from `namespaceUri`  to a kube cluster URI and compare it against `kubeClusterUri`.

On top of that, since there's no `routing.ensureKubeUri`, you have to manually call `routing.parseKubeResourceNamespaceUri` and `routing.getKubeUri` _and then_ you have to assing the result to a variable – otherwise the code would appear too busy.

This poses a problem – how are you going to name that intermediate variable? `namespaceKubeClusterUri`? This make it seem like it's a URI for a namespace. `kubeClusterUri` is already taken. `kubeClusterUriFromNamespaceUri` is a bit lengthy.

If you had `routing.ensureKubeUri`, you could write this:

```typescript
        if (routing.ensureKubeUri(namespaceUri) != kubeClusterUri) {
          throw new Error(
            'Only namespace belonging to the same requested kube cluster can be updated'
          );
        }
```

This could be taken one step further and we could add `routing.belongsToKube` (similar to `routing.belongsToProfile`:

```typescript
        if (routing.belongsToKube(kubeClusterUri, namespaceUri) {
          throw new Error(
            'Only namespace belonging to the same requested kube cluster can be updated'
          );
        }
```

Reply:
> > we could drop kubeClusterUri as an arg of this specific method of AccessRequestsService
> 
> hrm... but what if a namespace was cleared? i would still need to know the kube cluster URI to get to the namespaces b/c my list of namespaces will be empty:

Good point, I forgot that this might be the case.

Reply:
I like the idea of `routing.belongsToKube(kubeClusterUri, namespaceUri)`.